au/apply_magnitude_test.cc

```diff
@@ -128,5 +128,306 @@ TEST(ApplyMagnitude, MultipliesSingleNumberForIrrationalMagnitudeOnFloatingPoint
     EXPECT_THAT(apply_magnitude(2.0f, PI), SameTypeAndValue(2.0f * static_cast<float>(M_PI)));
 }
 
+TEST(WouldOverflow, HasCorrectBoundariesForIntegerMultiply) {
+    auto ONE_BILLION = pow<9>(mag<10>());
+
+    {
+        using ApplyOneBillionToI32 = ApplyMagnitudeT<int32_t, decltype(ONE_BILLION)>;
+
+        EXPECT_TRUE(ApplyOneBillionToI32::would_overflow(3));
+
+        EXPECT_FALSE(ApplyOneBillionToI32::would_overflow(2));
+        EXPECT_FALSE(ApplyOneBillionToI32::would_overflow(0));
+        EXPECT_FALSE(ApplyOneBillionToI32::would_overflow(-2));
+
+        EXPECT_TRUE(ApplyOneBillionToI32::would_overflow(-3));
+    }
+
+    {
+        using ApplyOneBillionToU8 = ApplyMagnitudeT<uint8_t, decltype(ONE_BILLION)>;
+
+        EXPECT_TRUE(ApplyOneBillionToU8::would_overflow(1));
+
+        EXPECT_FALSE(ApplyOneBillionToU8::would_overflow(0));
+    }
+
+    {
+        using ApplyOneBillionToF = ApplyMagnitudeT<float, decltype(ONE_BILLION)>;
+
+        EXPECT_TRUE(ApplyOneBillionToF::would_overflow(3.403e29f));
```

Comment:
Is there any significance to this value or was it found emperically?

Reply:
It is one one-billionth of the maximum `float` value, which I looked up.

(Well, slightly _above_ one one-billionth of that value.)